src/translations/maps/queuedPhysicalIndexToValueMap.js

```diff
@@ -0,0 +1,107 @@
+
+/**
+ * Map for storing mappings from an physical index to a value. Some values are stored in a certain order.
+ *
+ * Does not update stored values on remove/add row or column action. Otherwise, queue of indexes related to ordered
+ * values is updated after such changes.
+ */
+class QueuedPhysicalIndexToValueMap extends IndexMap {
+  /**
+   * Some values are stored in a certain order. Queue of indexes represent indexes related to ordered values.
+   *
+   * @type {Array<number>}
+   */
+  queueOfIndexes = []
+
+  /**
+   * Add values to list and reorganize. It updates queue of indexes related to ordered values.
+   *
+   * @private
+   * @param {number} insertionIndex Position inside the list.
+   * @param {Array} insertedIndexes List of inserted indexes.
+   */
+  insert(insertionIndex, insertedIndexes) {
+    this.indexedValues = getListWithInsertedItems(
+      this.indexedValues,
+      insertionIndex,
+      insertedIndexes,
+      this.initValueOrFn
+    );
+    this.queueOfIndexes = getIncreasedIndexes(this.queueOfIndexes, insertedIndexes);
+
+    super.insert(insertionIndex, insertedIndexes);
+  }
+
+  /**
+   * Remove values from the list and reorganize. It updates queue of indexes related to ordered values.
+   *
+   * @private
+   * @param {Array} removedIndexes List of removed indexes.
+   */
+  remove(removedIndexes) {
+    this.indexedValues = getListWithRemovedItems(this.indexedValues, removedIndexes);
+    this.queueOfIndexes = getListWithoutIndexes(this.queueOfIndexes, removedIndexes);
+    this.queueOfIndexes = getDecreasedIndexes(this.queueOfIndexes, removedIndexes);
+
+    super.remove(removedIndexes);
+  }
+
+  /**
+   * Set value at index and add it to the queue of values. Non-default values are stored in a certain order.
+   *
+   * Note: Queued value will be added at the end of the queue.
+   *
+   * @param {number} index The index.
+   * @param {*} value The value to save.
+   */
+  setValueAtIndex(index, value) {
+    super.setValueAtIndex(index, value);
+
+    this.queueOfIndexes.push(index);
+  }
+
+  /**
+   * Remove every queued value.
+   *
+   * Note: Please keep in mind that clear for the rest of the map won't be executed.
+   */
+  clear() {
+    if (isFunction(this.initValueOrFn)) {
+      this.queueOfIndexes.forEach((physicalIndex) => {
+        super.setValueAtIndex(physicalIndex, this.initValueOrFn(physicalIndex));
+      });
+
+    } else {
+      this.queueOfIndexes.forEach((physicalIndex) => {
+        super.setValueAtIndex(physicalIndex, this.initValueOrFn);
+      });
+    }
+
+    this.queueOfIndexes = [];
+  }
+
+  /**
+   * Get every queued value.
+   *
+   * @returns {Array}
+   */
+  getEntries() {
+    return this.queueOfIndexes.map(physicalIndex => [physicalIndex, this.getValueAtIndex(physicalIndex)]);
+  }
+
+  /**
+   * Get sequence of indexes related to values which have been queued.
+   *
+   * @returns {Array}
+   */
+  getIndexesQueue() {
+    return this.queueOfIndexes;
+  }
```

Comment:
I understand the idea, but I'm afraid that we shouldn't focus mainly on the possibility of redefining the map. The map provides new functionality as part of a **new structure**.

If it introduces new features thus it may have a new API for newly introduced parts of the functionality. This feature introduces information that **just some indexes** have non-default values and those **entries** are arranged in the order. That's why we introduced `getEntries` without overriding the old `getValues` method. Consistently I can't see a problem with introducing `getIndexesQueue` method (somehow it will be called finally).

Some methods can't be overridden as features of the "base" Index Map still may be used and I'm afraid that it should be still provided. The `getLength` method still represents the full size of the Index Map (also with "empty" values).

I'm concerned whether `clear` method should be overridden 🤔 .

Reply:
If we're going to use it as a LinkedMap, then the length (and therefore `getLength` value) is the number of linked elements. We have no interest in the empty values in this case. IMO `getValues` may be overwritten as well. Why would we need it in the current form if were interested in `queueOfIndexes` only?

Reply:
Hmm... I wrote [here](https://github.com/handsontable/handsontable/pull/7276#discussion_r499690664) that we create something unusual. From my perspective it look like:

![Untitled Diagram (11)](https://user-images.githubusercontent.com/141330/95203505-24204f80-07e3-11eb-9fd4-05574e72f9bd.jpg)

It's kind of hybrid where some values are just in the hash map, and part of them (it is important that it does not apply to all of them) are stored inside linked list.

Do we consider override of `getValueAtIndex` method? Now, we don't. Why? Because we are not dealing only a linked list (or precisely, Linked HashSet).

Reply:
Not that unusual ;) Look: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

The Map object has `Map.prototype.get(key)` that will return the value from the "Hash Set", so if the key does not exist it will return `undefined` just as our `getValueAtIndex` will return `null`.

At the same time Map object has `Map.prototype.values()` method that will return ONLY the values from the "Linked List". Moreover, it says "values for each element in the Map object **in insertion order**." so basically Javascript Map object is the Linked Hash Map.

IMO looks just like what you draw.

I agree with @budnix that we miss something from "Linked" structure. It's the iterator, you can't get `next()` linked item, just all `entries()` in the correct order. Something we can add later. Ideally, this IndexMap could be internally based on `Map.prototype`, not an `Array.prototype` so we have all of those out of the box and we would not need additional `queueOfIndexes` to store the order, JS does that for us.

Reply:
Okay, it convinces me. I can see that I incorrectly mixed current implementation with independent modeling of being. I will do my best to refactor it. Thanks 👍